enterprise/server/atime_updater/atime_updater.go

```diff
@@ -0,0 +1,239 @@
+	atimeUpdaterBatchUpdateInterval = flag.Duration("cache_proxy.remote_atime_update_interval", 30*time.Second, "The time interval to wait between sending access time updates to the remote cache.")
+	atimeUpdaterRpcTimeout          = flag.Duration("cache_proxy.remote_atime_rpc_timeout", 1*time.Second, "RPC timeout to use when updating blob access times in the remote cache.")
+)
+
+// A pending batch of atime updates (basically a FindMissingBlobsRequest).
+// Stored as a struct so digest keys can be stored in a set for de-duping.
+type atimeUpdate struct {
+	instanceName   string
+	digests        map[digest.Key]struct{} // stored as a set for de-duping.
+	digestFunction repb.DigestFunction_Value
+}
+
+func (u *atimeUpdate) toProto() *repb.FindMissingBlobsRequest {
+	req := repb.FindMissingBlobsRequest{
+		InstanceName:   u.instanceName,
+		BlobDigests:    make([]*repb.Digest, len(u.digests)),
+		DigestFunction: u.digestFunction,
+	}
+	i := 0
+	for d := range u.digests {
+		req.BlobDigests[i] = d.ToDigest()
+		i++
+	}
+	return &req
+}
+
+// The set of pending access time updates to send per group. There should be at
+// most one of these per group, but each contains many atimeUpdate{}s, as
+// there is one per (groupID, instance-name, digest-function) tuple. However,
+// the update will only keep one such atimeUpdate{} and if it gets too big will
+// discard additional digests until it's flushed.
+type atimeUpdates struct {
+	mu      sync.Mutex // protects jwt, updates, and atimeUpdate.digests.
+	jwt     string
+	updates []*atimeUpdate
+}
+
+type atimeUpdater struct {
+	authenticator interfaces.Authenticator
+
+	mu      sync.Mutex               // protects updates
+	updates map[string]*atimeUpdates // pending updates keyed by groupID.
+
+	maxDigestsPerUpdate int
+	maxUpdatesPerGroup  int
+	updateInterval      time.Duration
+	rpcTimeout          time.Duration
+
+	remote repb.ContentAddressableStorageClient
+}
+
+type Opts struct {
+	TickerForTesting <-chan time.Time
+}
+
+func Register(env *real_environment.RealEnv, opts Opts) error {
+	authenticator := env.GetAuthenticator()
+	if authenticator == nil {
+		return fmt.Errorf("An Authenticator is required to enable the AtimeUpdater.")
+	}
+	remote := env.GetContentAddressableStorageClient()
+	if remote == nil {
+		return fmt.Errorf("A remote CAS client is required to enable the AtimeUpdater.")
+	}
+	updater := atimeUpdater{
+		authenticator:       authenticator,
+		updates:             make(map[string]*atimeUpdates),
+		maxDigestsPerUpdate: *atimeUpdaterMaxDigestsPerUpdate,
+		maxUpdatesPerGroup:  *atimeUpdaterMaxUpdatesPerGroup,
+		updateInterval:      *atimeUpdaterBatchUpdateInterval,
+		rpcTimeout:          *atimeUpdaterRpcTimeout,
+		remote:              remote,
+	}
+	go updater.start(opts)
```

Comment:
Doesn't need to be in this PR, but I would like to see a best-effort attempt to drain the pending atime updates when shutdown is trigerrede.

Reply:
Good call, added a TODO